File: PathFinding.java
import java.util.*;

public class PathFinding {
  public int minTurns(String[] board) {
    int m = board.length;
    int n = board[0].length();
    int aPos = 0;
    int bPos = 0;
    for (int i = 0; i < m; ++i) {
      for (int j = 0; j < n; ++j) {
        if (board[i].charAt(j) == 'A') {
          aPos = (i << 16) | j;
        } else if (board[i].charAt(j) == 'B') {
          bPos = (i << 16) | j;
        }
      }
    }
    long initState = ((long)aPos << 32) | bPos;
    long targetState = ((long)bPos << 32) | aPos;
    //System.out.printf("initState = 0x%x, targetState = 0x%x\n", initState, targetState);
    Queue<Long> queue = new ArrayDeque<Long>();
    HashSet<Long> hitSet = new HashSet<Long>();
    int turn = 0;
    queue.add(initState);
    hitSet.add(initState);
    boolean found = false;
    while (!queue.isEmpty()) {
      int size = queue.size();
      while (size-- > 0) {
        long state = queue.remove();
        if (state == targetState) {
          found = true;
          break;
        }
        aPos = (int)(state >> 32);
        bPos = (int)(state & 0xffffffff);
        int aRow = aPos >> 16;
        int aCol = aPos & 0xffff;
        int bRow = bPos >> 16;
        int bCol = bPos & 0xffff;
        int[] dr = new int[]{0, 0, 0, -1, 1, -1, 1, -1, 1};
        int[] dc = new int[]{0, -1, 1, 0, 0, -1, 1, 1, -1};
        for (int i = 0; i < dr.length; i++) {
          for (int j = 0; j < dr.length; j++) {
            int aNextRow = aRow + dr[i];
            int aNextCol = aCol + dc[i];
            int bNextRow = bRow + dr[j];
            int bNextCol = bCol + dc[j];
            if (aNextRow >= 0 && aNextRow < m && aNextCol >= 0 && aNextCol < n &&
                bNextRow >= 0 && bNextRow < m && bNextCol >= 0 && bNextCol < n &&
                board[aNextRow].charAt(aNextCol) != 'X' && board[bNextRow].charAt(bNextCol) != 'X' &&
                !(aNextRow == bRow && aNextCol == bCol && bNextRow == aRow && bNextCol == aCol) &&
                !(aNextRow == bNextRow && aNextCol == bNextCol)) {
              long nextState = ((long)((aNextRow << 16) | aNextCol) << 32) | ((bNextRow << 16) | bNextCol);
              if (!hitSet.contains(nextState)) {
                //System.out.printf("state = 0x%x\n", nextState);
                queue.add(nextState);
                hitSet.add(nextState);
              }
            }
          }
        }
      }
      if (found) {
        break;
      }
      turn++;
    }
    return (found ? turn : -1);
  }
}

File: Pathfinding.py
class Pathfinding:
    def getDirections(self, x, y):
        if x > 0 and y > 0:
            if x % 2 == 0 or y % 2 == 0:
                return x + y
            else:
                return x + y + 2
        elif x > 0 and y == 0:
            return x + y
        elif x > 0 and y < 0:
            if x % 2 == 1 or y % 2 == 0:
                return x - y
            else:
                return x - y + 2
        elif x == 0 and y > 0:
            return x + y
        elif x == 0 and y == 0:
            return x + y
        elif x == 0 and y < 0:
            if y % 2 == 1:
                return x - y + 2
            else:
                return x - y + 4
        elif x < 0 and y > 0:
            if y % 2 == 1 or x % 2 == 0:
                return y - x
            else:
                return y - x + 2
        elif x < 0 and y == 0:
            if x % 2 == 1:
                return y - x + 2
            else:
                return y - x + 4
        elif x < 0 and y < 0:
            if x % 2 == 1 or y % 2 == 1:
                return 2 - x - y
            else:
                return 4 - x - y
